internal/guest/storage/scsi/scsi.go

```diff
@@ -175,27 +218,45 @@ func Mount(
 			}
 		}
 		source = encryptedSource
-		mountType = "xfs"
 	}
 
-	for {
-		if err := unixMount(source, target, mountType, flags, data); err != nil {
-			// The `source` found by GetDevicePath can take some time
-			// before its actually available under `/dev/sd*`. Retry while we
-			// wait for `source` to show up.
-			if errors.Is(err, unix.ENOENT) || errors.Is(err, unix.ENXIO) {
-				select {
-				case <-ctx.Done():
-					log.G(ctx).Warnf("mount system call failed with %s, context timed out while retrying", err)
+	// support mounting any filesystem by checking the device fs present
+	// and using that as the mountType unless `Filesystem` was given
+	deviceFS, err := _getDeviceFsType(source)
+	if err != nil {
+		return err
+	}
+	mountType := deviceFS
+	if config.Filesystem != "" {
+		mountType = config.Filesystem
+	}
+
+	// if EnsureFilesystem is set, then we need to check if the device has the
+	// correct filesystem configured on it. If it does not, format the device
+	// with the corect filesystem. Right now, we only support formatting xfs
+	// and ext4.
+	if config.EnsureFilesystem {
+		if deviceFS != config.Filesystem {
+			// re-format device to the correct fs
+			switch config.Filesystem {
+			case "ext4":
+				if err := ext4Format(ctx, source); err != nil {
 					return err
-				default:
-					time.Sleep(10 * time.Millisecond)
-					continue
 				}
+			case "xfs":
+				if err := xfsFormat(source); err != nil {
```

Comment:
in case of an encrypted scratch, `source` will be a dm-crypt target here and we don't want to format it. maybe L223-L254 should be under `else` of `if config.Encrypted`

Reply:
I added some code to not go into the `EnsureFilesystem` block if `config.Encrypted` is set. Does that seem good to you?

Reply:
yeah, sounds reasonable to me.